nixos/system-path: convert environment.systemPackages to an attrset

r-vdp opened the pull request:
## Description of changes

Lists are pretty bad for overriding, this PR proposes to use an attribute set for `environment.systemPackages` so that we can more easily disable packages or override the derivation for a package.

In order not to break the current API, we use the `coercedTo` type so that we can still accept a list of derivations and we convert this list into an attribute set. This should therefore be a non-breaking change.

We also allow for a shorthand syntax:
```nix
environment.systemPackages = {
  inherit (pkgs) acl;
};
```
Which gets coerced into
```nix
environment.systemPackages = {
  acl = {
    enable = true;
    package = pkgs.acl;
    outputs = {
      bin = true;
      man = true;
    };
  };
};
```
where the outputs were taken from `meta.outputsToInstall` and can be individually disabled (hence the attrset that maps every output to a boolean).

The nixos test gives a good overview of what the syntax would look like.

@roberth @infinisil @nikstur : I'd love to get your thoughts on this.

nikstur commented:
You should mark this as a draft if you're still working on it.

R-VdP commented:
Right, I removed the WIP label and converted to draft for now.

I'd want to do a bit more tidying up before merging, but before spending more time on this I'd first like to know that people agree with the approach.

nikstur commented:
I think the approach is really good. I've been discussing with some people that the systemPackages should be an overridable attrset but never acted on it, so I'm happy that someone else picked it up.

 The attrset way of declaring a package is a little verbose, but I'd assume that most people are just using the list way, so its fine.

infinisil commented on nixos/modules/config/system-path.nix:
```diff
@@ -6,6 +6,34 @@
+    in
+    ps: lib.listToAttrs (map mkEntry ps);
+
+  packagesByPrio = types.coercedTo
+    (types.listOf types.packageByPrio)
+    pkgsToAttrs
+    (types.attrsOf (types.submodule ( { lib, ... }: {
```
If used for `environment.systemPackages`, this is not entirely backwards compatible, because the result of this option is now an attribute set instead of a list, so any module making use of this option will break.

For `environment.systemPackages` the coercion should go the other way instead, which then also removes the need for the `pkgsToAttrs` function.

R-VdP replied:
But if we coerce in the other direction, then we lose all the benefit, since then we still end up with a list and we are in the same situation as today.

In the whole of nixpkgs, there was only one place where the value of `environment.systemPackages` is used, but it is being set from everywhere. And I also expect to see the same in peoples configs, that people just set packages but don't use the resulting value themselves.

So I think what we want is that the actual value of the option is always an attrset, so that we can have the overrides and such, but maintain the possibility of setting the option using a list, to maintain backwards compat, which is what the current type does.

R-VdP replied:
Yeah, I still doubt that anyone directly consumes this option since `system.path` is almost always what you'd want instead, but I guess it wouldn't hurt to add `apply`.

I'd like to avoid introducing a new option since it makes things even more complex, and might require people to understand a lot of internals to understand how to override a package on the system path.

R-VdP replied:
I've added an apply, this seems to work nicely, thanks for the suggestion!

infinisil commented on lib/types.nix:
```diff
@@ -475,6 +475,47 @@ rec {
 
+    # This type is like `package` but it gets merged based on the
+    # `meta.priority` attribute.
+    packageByPrio =
```
How does this interact with `mkForce` and co.? Those should definitely also work.

Actually `meta.priority` is kind of just a mechanism to get around not having a proper override mechanism before. But now that we do with `mkForce` and co., it shouldn't be necessary anymore.

I'd be in favor of ignoring `meta.priority`'s on this level, or even complaining if they're set (for attribute sets), referring people to `mkForce` and co. instead.

R-VdP replied:
Yeah `mkForce` works (see the test case). The priority is currently only used when merging, so only if the same package is set multiple time with the same module-level priority.

We could try to remove it, but it would take a bit more effort to re-encode the current meta priorities as module-system priorities.

roberth replied:
`meta.priority` is about contents, whereas module level priorities are about replacing actual packages. Those are distinct solutions to distinct problems.
Two interesting cases:
- Remove a command by replacing a package. This can only be done by overriding the `.package` with the module system.
- Resolve a conflict between two packages that both want to create some file. Use `meta.priority` to pick the best one. You probably don't want to delete the other files contributed by the lower package.

R-VdP replied:
So do I understand correctly that we are OK with the changes as proposed in this PR currently, or do we need any changes still?

R-VdP replied:
This morning I realised that the whole priority concept is actually not relevant here. We don't need a new type if we just change the already existing `package` type to use `mergeEqualOption` (I don't see a reason for using `mergeOneOption` there) so that if we assign the same derivation twice, it will not cause a conflict.

The priority concept is only used while building the final environment containing all system packages and is used to resolve conflicts when the same file is present in multiple derivations.

R-VdP commented on nixos/modules/config/system-path.nix:
```diff
@@ -6,6 +6,34 @@
+  packagesByPrio = types.coercedTo
+    (types.listOf types.packageByPrio)
+    pkgsToAttrs
+    (types.attrsOf (types.submodule ( { lib, ... }: {
+      options = {
+        enable = lib.mkEnableOption "the package";
+        package = lib.mkOption {
+          type = types.packageByPrio;
```
Hmm, and you would then do something like `environment.systemPackages.hello = null;` to remove a package.

This does not work well for multi-output packages though. How would you include both `shadow` and `shadow.su`?

infinisil replied:
Oh, that makes me think that outputs should be an explicit option in of itself, maybe let's  have both syntaxes for packages like this:

```
environment.systemPackages = {
  inherit (pkgs) hello;
  # The above gets coerced to the below
  hello = {
    enable = true;
    package = pkgs.hello;
    outputs.out = true;
  };
```

Where the `outputs` default to something derived from https://github.com/NixOS/nixpkgs/blob/3eab452b7a9e4068b6ff7cbd10f1b96dea4fa962/pkgs/build-support/buildenv/default.nix#L64-L67

Then one can change the outputs that get installed by setting e.g. `environment.systemPackages.hello.outputs.bin = true`

R-VdP replied:
That looks nice indeed.

But we'd need two levels of coercions then, something like
```
type = coercedTo (listOf packages) (...) (
  coercedTo (attrsOf package) (...) (submodule {
    options = {
      enable = mkEnableOption "...";
      package = mkOption { ... };
      outputs = mkOption {
        type = attrsOf bool;
      };
    };
  })
);
```

I'd need to experiment a bit with that to see how well it works.

R-VdP replied:
This is done now.

R-VdP commented:
@nikstur after implementing the suggestions from @infinisil, the way to declare a package using the attrset approach became a lot more convenient, as you can see in the test case. We can deduce the outputs automatically now.

roberth commented on nixos/modules/config/system-path.nix:
```diff
@@ -7,6 +7,97 @@ with lib;
+    types.coercedTo
+      (types.listOf types.package)
+      pkgsToAttrs
+      (types.coercedTo
```
It seems that this could be `attrsOf (coercedTo` instead. That would simplify the rest of the type and iiuc behave better for mixed definitions (e.g. `systemPackages = { foo = pkgs.foo; bar = { package = pkgs.bar }; }`, which seems like a good test case as well)

R-VdP replied:
I tried this, but the big downside is that we cannot control the names anymore in that case, and so we get a difference in behaviour between `environment.systemPackages = [ pkgs.hello ];` and `environment.systemPackages = { inherit (pkgs) hello; };`.

In the first case we get `"${pkgs.hello.name}"` as attribute name (since not every derivation has a pname), but in the second case we get `"hello"`, and so the way to override this package would depend on how it was originally put into `systemPackages`, which seems unnecessarily brittle.

So we need a naming that can be derived from the derivation since in the case of a list we don't have anything else to derive the name from when we convert the list into an attr set. And ideally we would then make sure that the two other ways of defining `systemPackages` use the exact same naming scheme, so that things are stable and predictable.

What do you think?

roberth replied:
I might have been unclear with my suggestion. I meant to construct the type as

```nix
coercedTo
  (listOf package)
  pkgsToAttrs
  (attrsOf t)
```

where `t` is the logic below, except treating a single package at a time.

The issue you're describing seems to be about `pkgsToAttrs`, which I didn't mean to touch.

roberth replied:
Regarding the `pkg.name` problem, I'd be happy to give them really bad names, such as the drv hash, so that overridability is only possible when the original definition has an explicit attribute name.

Suppose NixOS switches to `sudo-rs` by default, while using the list syntax. That would break all configs that have an override on `sudo`.
I'd say lists are bad, and explicit is good, so let's make people use the attrset syntax if they want to use this overriding feature.

R-VdP replied:
> I might have been unclear with my suggestion. I meant to construct the type as
> 
> ```nix
> coercedTo
>   (listOf package)
>   pkgsToAttrs
>   (attrsOf t)
> ```
> 
> where `t` is the logic below, except treating a single package at a time.
> 
> The issue you're describing seems to be about `pkgsToAttrs`, which I didn't mean to touch.

I think we're talking about the same thing, but what you're proposing doesn't really work well.

So what you're proposing is something like this:
```nix
coercedTo
  (listOf package)
  pkgsToAttrs
  (attrsOf (coercedTo
    package
    packageToStructuredAttrs
    (submodule { ... })
  ))
```

The problem is that in this case when we write
```nix
environment.systemPackages = { inherit (pkgs) hello; };
```
We get as attribute name `"hello"`, but when we convert from the list of packages, we get the full `name` attribute (or the drv hash or whatever we choose for it, but usually not the attribute name of the package in `pkgs`).
And so we cannot override between these two because the attribute names are inconsistent. And since our coercion function is only action on the values of the attribute sets, and not on the sets themselves, we cannot change the attribute names.

So in this case we cannot have the same attribute name in the final attribute set for
`systemPackages = [ pkgs.hello ];` and `systemPackages = { inherit (pkgs) hello; }`.

The big difference with the current implementation, is that when we have
```nix
coercedTo
  (listOf package)
  pkgsToAttrs
  (coercedTo
    (attrsOf package)
    packagesToStructuredAttrs
    (attrsOf (submodule { ... }))
  )
```
we can change the attribute names in `packagesToStructuredAttrs`, since we are converting one attrset into another.

R-VdP replied:
Agreed, but then we need to have the type that I'm currently proposing, as `{ inherit (pkgs) hello; }` uses the attribute name (usually `pname`) and with the type you proposed we cannot rename this to a more stable name.

roberth replied:
> as `{ inherit (pkgs) hello; }` uses the attribute name (usually `pname`)

Why would it do that? This is `{ hello = pkgs.hello; }` where the first `hello` is provided by the module author. That's as stable as it gets. (I'd recommend against `inherit` syntax here, but that's besides the point)

> we cannot rename

We don't need to rename anything. Let the module author write the correct name. That's the only way we don't accidentally use bad data.

R-VdP replied:
It's not the module author, it's the attribute in `pkgs`.

The point is that without renaming, we cannot get a stable interface for overriding, since the name in the final attrset will depend on how the package was added, which means that stuff breaks when people migrate from the list syntax to the attrset syntax.

R-VdP replied:
Ok, I see where you are coming from. You don't want to enable overriding of packages added through the list syntax.
I see arguments for both approaches, but I don't care enough to have this PR blocked on this, so I implemented it in the way you suggested now.

Should we then also deprecate the list syntax and print a warning when it's used, so that we can migrate to the new syntax that allows for overriding?

roberth replied:
We could deprecate it when 23.05 is EOL, so that third party modules that target the supported releases aren't stuck between a warning and an error. This can be achieved with `warnIf (lib.isInOldestRelease 2311)`.

roberth commented on nixos/modules/config/system-path.nix:
```diff
@@ -7,6 +7,97 @@ with lib;
+          config = {
+            # We add the default outputs of the package by default.
+            # If a specific output has been selected, we only add that one.
+            outputs = mkOutputs (outputsFor config.package);
```
Should this be a `mkDefault`?

R-VdP replied:
I don't think so because then we cannot add an output to an already included package, since it would overwrite the previously included outputs if they have lower priorities.

I wrote it like this so that we add on to the default outputs by default, if someone really want to override, they can do so with `mkForce`.

R-VdP commented on nixos/modules/config/system-path.nix:
```diff
@@ -7,6 +7,97 @@ with lib;
+    let
+      enabled = lib.filterAttrs (_: p: p.enable) ps;
+      packageToOutputs = p:
+        mapAttrsToList (output: _: p.package.${output}) (lib.filterAttrs (_: lib.id) p.outputs);
```
Ok, it doesn't make the code simpler, but I agree on not duplicating this logic.

roberth commented on lib/types.nix:
```diff
@@ -469,7 +469,7 @@ rec {
       check = x: isDerivation x || isStorePath x;
       merge = loc: defs:
-        let res = mergeOneOption loc defs;
+        let res = mergeEqualOption loc defs;
```
This seems risky, because equality on derivation-like things is not accurate. We don't want to pick an arbitrary package when that happens.

```
> zlib == zlib // { yikes = true; }
true
```

R-VdP replied:
Hmm that's a good point. I think we can compare on the `outPath` then?

The issue that we're trying to avoid here is that the same package can be added multiple times with the list syntax, and we don't want to get an error from the module system in that case when they are exactly the same, in that case we can just deduplicate and continue.

R-VdP replied:
The does mean that things like `passthru` are not taken into account in the deduplication though.

rnhmjoj replied:
> That would mean that things like passthru are not taken into account in the deduplication though.

This could be a problem: I know several packages that provide plugins this way and you may want to override the package to add your own plugin, for example.

R-VdP replied:
But the final package with the plugin added would still have a different `outPath` then, wouldn't it?

The passthru itself is not part of the drv, so it can only provide things at eval time, but it cannot affect the resulting build output.

roberth replied:
I don't think it's necessary for this change to be part of this PR.

Final thought: probably an appropriate equality for packages depends on the use case, so we might want to do this not in `types.nix`, but rather make exceptions where in concrete cases where we can reason about which definitions are equivalent given how the option is used.

infinisil replied:
See https://github.com/NixOS/nixpkgs/pull/223064

roberth commented:
I think the following definition leads to an unfortunate error message:

```nix
systemPackages = pkgs.hello;
```

`hello` is an attrset, so its package attributes will be treated as packages instead. I think we should detect this mistake in the type and report that it should be `systemPackages.hello = <pkg>`.

roberth commented:
This also needs a release note and an update to the option description to clarify what the attribute name is for.

NobbZ commented:
Maybe this is a very dumb question, but…

> Lists are pretty bad for overriding

Do you have any examples how they are bad for overriding, and how a set would be any better?

R-VdP commented:
> Maybe this is a very dumb question, but…
> 
> > Lists are pretty bad for overriding
> 
> Do you have any examples how they are bad for overriding, and how a set would be any better?

How would you go about overriding an existing package from systemPackages now? They don't have names that you can use to reference to them.

eclairevoyant commented:
> How would you go about overriding an existing package from systemPackages now? They don't have names that you can use to reference to them.

Depends how they are being added. If I'm adding to the list directly myself:
```nix
environment.systemPackages = [
  (pkgs._7zz.override {enableUnfree = true;})
];
```

If I'm using a module, then I hope that the module exposes a `.package` option where I can set such an overridden package.

roberth commented:
> > Lists are pretty bad for overriding

This is a general observation in the context of a single option. If all modules that define the option provide their own mechanisms to customize it, that would be fine. (And with `systemPackages` they usually do, but this was meant to be a general observation.)

> Do you have any examples how they are bad for overriding, and how a set would be any better?

As mentioned, with attrsets, the attribute is an unambiguous key that you can use to do a `lib.mkForce`.
Lists also have keys - integers instead of strings - but these aren't unambiguous in a module system context, because of definition merging. If you have two modules that each contribute definitions to the option, the overriding module would need global knowledge about all modules to figure out the right key to override. Furthermore, the `listOf t` type has a merge operator that isn't commutative, so it would also couple to the order of `imports` and the import resolution algorithm (which currently performs strange reversals...)

infinisil commented on nixos/modules/config/system-path.nix:
```diff
@@ -7,6 +7,118 @@ with lib;
+    let
+      inherit (lib) types;
+
+      outputsFor = drv:
+        if drv.outputSpecified or false
+        then [ drv.outputName ]
+        else ((drv.meta or {}).outputsToInstall or [ (drv.outputName or "out") ]);
```
The same logic is in `pkgs.buildEnv`: https://github.com/NixOS/nixpkgs/blob/670182de729294664672addedf6ecef2095037c5/pkgs/build-support/buildenv/default.nix#L60-L70

Would be good to deduplicate this, probably into its own function in `lib.derivations.*`

infinisil commented on nixos/modules/config/system-path.nix:
```diff
@@ -7,6 +7,118 @@ with lib;
+
+      drvToAttrs = drv:
+        let
+          defaultOutput = (drv.${(lib.head (drv.outputs or [ "out" ]))}
+            or (throw "something is up with the outputs of this derivation: ${drv}")) // {
```
Same here, this should be deduplicated with a lib function. We can also more easily test the code that way.

infinisil commented on nixos/modules/config/system-path.nix:
```diff
@@ -7,6 +7,118 @@ with lib;
+          outputs = mkOutputs (outputsFor drv);
+        };
+
+      drvToAttrsFolder = acc: drv: lib.recursiveUpdate acc {
+        "${builtins.unsafeDiscardStringContext drv.drvPath}" = drvToAttrs drv;
+      };
+
+      pkgsToAttrs = lib.foldl' drvToAttrsFolder {};
```
`recursiveUpdate` is dangerous because it silently overrides values (which indeed can happen here). It shouldn't be necessary if you know the structure. In this case something like this should work:

```nix
pkgsToAttrs = list:
  lib.mapAttrs (name: values:
    if length values > 1 then
      ?
    else
      head values
  ) (lib.groupBy (drv:
    builtins.unsafeDiscardStringContext drv.drvPath
  ) list
```

infinisil replied:
Alternatively to @roberth's idea of using the `drvPath` as the attribute name, we could also generate attribute names from the list index, such that you can't even get a stable index to override. This would at least avoid the somewhat ugly `unsafeDiscardStringContext` and attribute names.

infinisil commented on nixos/modules/config/system-path.nix:
```diff
@@ -7,6 +7,118 @@ with lib;
+
+      pkgsToAttrs = lib.foldl' drvToAttrsFolder {};
+
+      packageByOutPath =
```
I see you commented in the [other PR](https://github.com/NixOS/nixpkgs/pull/223064) already, but this change should really be done separately

infinisil commented on nixos/modules/config/system-path.nix:
```diff
@@ -7,6 +7,118 @@ with lib;
+        let
+          outputsToInstall = lib.mapAttrsToList lib.const (lib.filterAttrs (_: lib.id) p.outputs);
+        in
+        lib.optional (lib.length outputsToInstall > 0) (p.package // {
+          meta = (p.package.meta or {}) // { inherit outputsToInstall; };
+        });
```
This also looks like something that should be in a `lib` function.

infinisil commented on nixos/modules/config/system-path.nix:
```diff
@@ -7,6 +7,118 @@ with lib;
+          meta = (p.package.meta or {}) // { inherit outputsToInstall; };
+        });
+    in
+    lib.flatten (lib.mapAttrsToList (_: setOutputs) enabled);
```
`flatten` should never be necessary when you know the structure. In this case the faster `concatLists` can be used.

infinisil commented on nixos/modules/config/system-path.nix:
```diff
@@ -143,7 +257,10 @@ in
 
   config = {
 
-    environment.systemPackages = requiredPackages ++ config.environment.defaultPackages;
+    environment.systemPackages = lib.mkMerge [
+      requiredPackages
+      config.environment.defaultPackages
+    ];
```
`defaultPackages` is exactly where this change shines, because the only reason `defaultPackages` was introduces is to be able to turn off packages installed by default! The default packages should definitely be changed to use the new attribute set definitions, so that its definitions can be overridden.

infinisil reviewed (changes requested):
Generally I think this is the right approach now! This should be split into separate PR's though.

roberth commented:
As a general note, I'd recommend to focus on the implementation first before polishing any new library functions, to avoid polishing the wrong functions.

r-vdp commented:
@roberth @infinisil it took me a bit of time to get back to this PR, but I think I have addressed all comments now.
Would you mind giving this another look?

Thanks!

r-vdp committed:
lib.derivations: init defaultOutput, outputsFor, setOutputsToInstall

r-vdp committed:
build-support/buildenv: use lib.derivations.outputsFor

r-vdp committed:
lib.types: init packageByOutPath

r-vdp committed:
system-path: convert environment.systemPackages to an attrset

This allows easily disabling or overriding a package.

r-vdp commented:
I pushed a fix for build-env which should get rid of all the rebuilds.

r-vdp committed:
lib.meta: init setPrioRecursively

r-vdp commented:
I checked several of the rebuilds and they are due to the ordering of the paths in the resulting list. This was in definition order before, and is in alphabetical order now, due to attrsets being sorted.

r-vdp committed:
nixos/system-path: correctly deal with packages with multiple outputs that have different priorities

infinisil reviewed (commented):
This PR currently adds 4 non-trivial, untested and undocumented `lib` functions. Either those should be done in smaller PRs (see also the new [lib PR guidelines](https://github.com/NixOS/nixpkgs/tree/master/lib#pr-guidelines)), or we shouldn't expose them publicly, as [hinted](https://github.com/NixOS/nixpkgs/pull/255086#issuecomment-1731146174) by Robert.

roberth commented on nixos/modules/config/system-path.nix:
```diff
@@ -6,50 +6,209 @@
+                      });
+                      default = outerConfig.package.${name}.meta.priority or null;
+                      defaultText = ''
+                        The meta.priority value of the package, or null if there is no priority assigned.
```
Could you describe the effect of `priority`?
E.g. what does a large value do? Do packages without a priority have a default priority?
I think you could link to https://nixos.org/manual/nixpkgs/stable/#var-meta-priority, but that doc isn't much better.
Maybe this? https://nixos.org/manual/nix/stable/command-ref/nix-env/install.html?highlight=priority#description

I think it'd be ok to copy some of this information, because
 - NixOS/nixpkgs duplicates this logic anyway
 - It's not going to change in Nix, because that'd break profiles for no good reason

roberth reviewed (commented):
Some clarifications needed

SuperSandro2000 commented:
I am sorry if that was already answered in some thread, I didn't read all of them but what happens to packages that where before duplicated by name in the list but not by the files they contain?

r-vdp commented:
> I am sorry if that was already answered in some thread, I didn't read all of them but what happens to packages that where before duplicated by name in the list but not by the files they contain?

What do you mean exactly by "the files they contain"? Packages added through the list syntax, end up in the attrset with their drv path as key, so unless they come from the same derivation, they would end up under different keys. If they do come from the same derivation, then their outputs will be combined in the list of outputs to install.

Does that answer your question?

SuperSandro2000 commented:
Yeah, I first thought they would be using the package variable as a key which isn't unique but the drv path must be unique. 👍🏼

r-vdp commented:
There's currently still an issue with `bash` being selected instead of  `bashInteractive`, I need to find the time to figure out why that is.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/question-about-defining-environment-systempackages-via-builtins-attrvalues-versus-direct-list/43410/3

nyabinary commented:
> There's currently still an issue with `bash` being selected instead of `bashInteractive`, I need to find the time to figure out why that is.

Any update on this? :3

RazYang commented:
Any progress?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/82

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/from-a-nixos-module-is-it-possible-to-delete-an-entry-from-an-attrset-defined-by-a-different-module/73289/17
